api/src/trace/SugaredTracer.ts

```diff
@@ -0,0 +1,232 @@
+   *      // do some async work
+   *     });
+   */
+  withSpan<F extends (span: Span) => ReturnType<F>>(
+    name: string,
+    fn: F,
+  ): ReturnType<F>
+  withSpan<F extends (span: Span) => ReturnType<F>>(
+    name: string,
+    options: SpanOptions,
+    fn: F,
+  ): ReturnType<F>
+  withSpan<F extends (span: Span) => ReturnType<F>>(
+    name: string,
+    options: SpanOptions,
+    context: Context,
+    fn: F,
+  ): ReturnType<F>
+  withSpan<F extends (span: Span) => ReturnType<F>>(
+    name: string,
+    options: SpanOptions,
+    context: Context,
+    fn: F,
+  ): ReturnType<F>
+  withSpan<F extends (span: Span) => ReturnType<F>>(
+    name: string,
+    options: SpanOptions,
+    context: Context,
+    onException: (e: Error, span: Span) => void,
+    fn: F,
+  ): ReturnType<F>
+  withSpan<F extends (span: Span) => ReturnType<F>>(
+    name: string,
+    arg2: SpanOptions | F,
+    arg3?: Context | F,
+    arg4?: ((e: Error, span: Span) => void) | F,
+    arg5?: F,
+  ): ReturnType<F> {
+    const {opts, ctx, onException, fn} = massageParams(arg2, arg3, arg4, arg5);
+
+    const span = this._tracer.startSpan(name, opts, ctx);
+    return handleFn(span, onException, fn) as ReturnType<F>;
+  }
+}
+
+/**
+ * Massages parameters of withSpan and withActiveSpan to allow signature overwrites
+ * @param arg
+ * @param arg2
+ * @param arg3
+ * @param arg4
+ */
+function massageParams<F extends (span: Span) => ReturnType<F>>(arg: SpanOptions | F,
+  arg2?: Context | F,
+  arg3?: ((e: Error, span: Span) => void) | F,
+  arg4?: F) {
+  let opts: SpanOptions | undefined;
+  let ctx: Context | undefined;
+  let onException = defaultOnException;
+  let fn: F;
+
+  if (arguments.length === 1) {
+    fn = arg as F;
+  } else if (arguments.length === 2) {
+    opts = arg as SpanOptions;
+    fn = arg2 as F;
+  } else if (arguments.length === 3) {
+    opts = arg as SpanOptions;
+    ctx = arg2 as Context;
+    fn = arg3 as F;
+  } else {
+    opts = arg as SpanOptions;
+    ctx = arg2 as Context;
+    onException = arg3 as (e: Error, span: Span) => void;
+    fn = arg4 as F;
+  }
+  opts = opts ?? {};
+  ctx = ctx ?? api.context.active();
+
+  return {opts, ctx, onException, fn};
+}
+
+/**
+ * Executes fn, returns results and runs onException in the case of exception to allow overwriting of error handling
+ * @param span
+ * @param onException
+ * @param fn
+ */
+function handleFn<F extends (span: Span) => ReturnType<F>>(span: Span, onException: ((e: Error, span: Span) => void), fn: F): ReturnType<F> {
+  try {
+    const ret = fn(span) as Promise<ReturnType<F>>;
+    // if fn is an async function attach a recordException and spanEnd callback to the promise
+    if (
+      typeof ret.then === 'function' &&
+      typeof ret.catch === 'function'
+    ){
+      return ret
+        .catch((e: Error) => {
+          onException(e, span);
+        })
+        .finally(span.end) as ReturnType<F>;
```

Comment:
```suggestion
        .finally(() => span.end) as ReturnType<F>;
```

There are no guarantees that `span.end` should be bound so it would be better to call `span.end` the proper receiver.